### ISSUE TITLE
Support graphs/trees/networks

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#187]
Need to add support for visualizing graph data. Need to integrate with various formats (dot, networkx, etc.), support things like force directed layouts, etc.

**Comment by bryevdv** [comment]
@JosephCottam any update on experiments with networkX?

Also, a potentially useful reference from @pzwang http://sigmajs.org/

**Comment by bixb0012** [comment]
Although this is already a [long-term](https://github.com/bokeh/bokeh/milestones/long-term) milestone,
+1

**Comment by hhuuggoo** [comment]
The other thing we could do is render an svg on top of the canvas and then allow a few d3 based renderers

**Comment by tonyfast** [comment]
While this issue is still pending.  Please see this [notebook example](http://nbviewer.ipython.org/gist/tonyfast/635714a234703d187452) that uses Bokeh to plot NetworkX graphs.

**Comment by franciscovargas** [comment]
I have a tiny example script which gives you ok looking graphs (its a binary tree) and I have not automated it attm to represent any adjecency list based graph but just using basic dict-list representations it is not too difficult to have this working I believe:
```python
from bokeh.plotting import show, figure
from bokeh.models import HoverTool, ColumnDataSource
from collections import OrderedDict


fig = figure(title='DDS Help Tree',
         tools= 'box_zoom,box_select,resize,reset,hover')
names = ["Gavin Gray", "Ewan Klein", "Francisco Vargas", "S3", "S1", "S2", "S4"]
info = ["TA", "Lectuer", "TA", "Student", "Student", "Student", "Student"]
fig.xgrid.grid_line_color = None
fig.ygrid.grid_line_color = None
fig.axis.major_label_text_font_size = '0pt'  
fig.axis.major_tick_line_color = None
fig.axis[0].ticker.num_minor_ticks = 0
fig.axis[1].ticker.num_minor_ticks = 0
fig.outline_line_color = "white"
fig.xaxis.axis_line_color = "white"
fig.yaxis.axis_line_color = "white"
fig.line(x=[1,3,5],y=[1,2,1])
fig.line(x=[0,1,2],y=[0,1,0])
fig.line(x=[4,5,6],y=[0,1,0])
source = ColumnDataSource(
    data=dict(
        xname=[1, 3, 5,6,0,2,4],
        yname=[1, 2, 1,0,0,0,0],
        name = names,
        info = info
    )
)
c1 = fig.circle(x='xname', y='yname',radius=0.29,color='#6495ED',source=source)
fig.annulus(x='xname', y='yname', inner_radius=0.29,
            outer_radius=0.295,source=source)
x1,y1 = (-0.1, -0.04)
fig.text(x=[-0.1, x1 + 2, x1 + 4, x1+6] , y=[y1, y1, y1 , y1] , text=["S1", "S2", "S3", "S4"])
fig.text(x=[x1 + 0.9, x1 + 5] , y=[y1 +1 , y1 +1] , text=["GG", "FV"])
fig.text(x=[x1 + 3] , y=[y1 +2] , text=["EK"])
hover = c1.select(dict(type=HoverTool))
hover.tooltips = OrderedDict([
    ('name', '@name'),
    ('info', '@info'),
])
show(fig)
```
And it looks like this (currently with some issues with hover tool):

![bokeh](https://cloud.githubusercontent.com/assets/5540172/9408802/922309aa-480d-11e5-83aa-7dacd9f51228.png)

**Comment by birdsarah** [comment]
You might also be interested in the network I've been working on with @gmarkall - http://nbviewer.ipython.org/github/gmarkall/wot/blob/master/wot/notebooks/Visualize%20the%20WOT.ipynb?flush_cache=True

![screenshot from 2015-08-21 14 39 42](https://cloud.githubusercontent.com/assets/1796208/9409528/26b9ca90-4813-11e5-8582-4b322dc8c853.png)

**Comment by ToferC** [comment]
Would love to see native graphing support in Bokeh.

**Comment by tdennist** [comment]
+1

**Comment by kingdynasty** [comment]
+1

**Comment by ngehman** [comment]
+1

**Comment by rnowling** [comment]
It would very nice if bokeh supported collapsable trees for visualizing things like file systems or hierarchical data.  Not sure if this would be same implementation as this issue or a new issue.

**Comment by datnamer** [comment]
Heres an interesting ggplot inspired take on a graph plotting api https://github.com/thomasp85/ggraph

**Comment by jainrp** [comment]
+1 Add support for graphing in bokeh, please!

**Comment by denfromufa** [comment]
+1

**Comment by rodac5** [comment]
+1

**Comment by marvelousmishig** [comment]
+1

**Comment by y-abe** [comment]
+1

**Comment by kingdynasty** [comment]
+1

**Comment by bryevdv** [comment]
If rendering graph data that is played out on the python side is sufficient, see e.g.

https://github.com/bokeh/bokeh/blob/master/examples/plotting/file/graphs.py

Adding dynamic graph layout to the JS side is another matter. To add it to the core library will require lots of deliberation, planning, testing. However, it would probably not be too difficult to write a custom extension that wraps e.g. the d3 graph layout library and connects it to Bokeh python. There's a new user's guide section on extending Bokeh that could be helpful:

http://bokeh.pydata.org/en/latest/docs/user_guide/extensions.html

We could also help anyone develop a custom extension by answering questions, etc. I'd suggest this is probably the quickest route for now, and the custom extension could be shared with the community which I am sure many people would appreciate. (Making it trivial to create and share extensions is on the short term roadmap, so that the community can move forward itself, and not have to "wait on us" for everything)

**Comment by abojchevski** [comment]
+1

**Comment by avyfain** [comment]
+1

**Comment by b3yang** [comment]
+1

**Comment by sumitbinnani** [comment]
+1

**Comment by tbrittoborges** [comment]
+1

**Comment by mbewley** [comment]
+1

**Comment by tgetachew18** [comment]
+1

**Comment by bryevdv** [comment]
Please, there is no use to continued +1 It does not add any information or help. I will repeat what I said earlier this year:

If rendering graph data that is layed out statically on the python side (e.g. by using NetworkX) is sufficient, see e.g.

https://github.com/bokeh/bokeh/blob/master/examples/plotting/file/graphs.py

which is possible as of 0.12.0.

Adding dynamic graph layout to the JS side is another matter. To add it to the core library will require lots of deliberation, planning, testing. However, it would probably not be too difficult to write a custom extension that wraps e.g. the d3 graph layout library (or another JS library) and connects it to Bokeh python. There's a new user's guide section on extending Bokeh that could be helpful:

http://bokeh.pydata.org/en/latest/docs/user_guide/extensions.html

We could also help anyone develop a custom extension by answering questions, etc. I'd suggest this is probably the quickest route for now, and the custom extension could be shared with the community which I am sure many people would appreciate. (Making it trivial to create and share extensions is on the short term roadmap, so that the community can move forward itself, and not have to "wait on us" for everything)

**Comment by bryevdv** [comment]
There is now supported work to add network/graph capability to Bokeh, for not-huge graph sizes. There is corresponding supported work to add complementary capability to Datashader for large graph sizes. Tenatively marking for `0.12.7`

**Comment by threecifanggen** [comment]
+1
